ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -1881,18 +2057,318 @@ void shutdown() throws InterruptedException {
+     * @param remoteReplicas {@link List}
+     * @param standByGroupId {@link int}
+     * @param groupId {@link int}
+     */
+    private void getActiveAndStandByNoProgressReplicas(List<RemoteReplicaInfo> activeReplicas,
+        List<RemoteReplicaInfo> standbyReplicasWithNoProgress, List<RemoteReplicaInfo> remoteReplicas,
+        int standByGroupId, int groupId) {
+      // this will contain all replicas which are available for iteration
+      Set<RemoteReplicaInfo> nonIteratingNonPendingRemoteReplicas = new LinkedHashSet<>(remoteReplicas);
+
+      // remove all throttled replicas
+      remoteReplicas.forEach(remoteReplicaInfo -> {
+        if (replicaToThrottledTill.containsKey(remoteReplicaInfo)
+            && replicaToThrottledTill.get(remoteReplicaInfo) > time.milliseconds()) {
+          nonIteratingNonPendingRemoteReplicas.remove(remoteReplicaInfo);
+          throttleCount.inc();
+        }
+      });
+
+      // remove all replicas of corresponding group id if group is in iteration
+      if (inflightGroups.containsKey(groupId)) {
+        logger.trace("Thread name: {} group {} is in iteration", threadName, groupId);
+        nonIteratingNonPendingRemoteReplicas.removeAll(inflightGroups.get(groupId).remoteReplicaInfos);
+      }
+
+      // remove all replicas of corresponding standby group if group is in iteration
+      if (inflightGroups.containsKey(standByGroupId)) {
+        logger.trace("Thread name: {} stand by group {} is in iteration", threadName,
+            standByGroupId);
+        nonIteratingNonPendingRemoteReplicas.removeAll(
+            inflightGroups.get(standByGroupId).remoteReplicaInfos);
+      }
+
+      // filter active and stand by replicas
+      filterRemoteReplicasToReplicate(new ArrayList<>(nonIteratingNonPendingRemoteReplicas), activeReplicas,
+          standbyReplicasWithNoProgress);
+    }
+
+
+    /**
+     * Iterates over {@link #groupIdToRemoteReplicaMap} all potential groups,
+     * tries to create active groups and adds new groups to {@link #inflightGroups}
+     * also adds replicas which have timed out on stand by
+     * to {@link #standByReplicaQueue} for being picked by standby groups
+     */
+    private void createActiveGroups(){
+      logger.trace("Thread name: {} Trying to create new active groups", threadName);
+      for (Map.Entry<Integer, List<RemoteReplicaInfo>> entry : groupIdToRemoteReplicaMap.entrySet()) {
+        int groupId = entry.getKey();
+        logger.trace("Thread name: {} Processing for group id {}", threadName, groupId);
+        List<RemoteReplicaInfo> remoteReplicas = entry.getValue();
+        DataNodeId remoteNode = remoteReplicas.get(0).getReplicaId().getDataNodeId();
+
+        // get stand by group id for the data node
+        int standByGroupId = standByGroupIds.get(remoteNode);
+
+        List<RemoteReplicaInfo> activeReplicas = new ArrayList<>();
+        List<RemoteReplicaInfo> standbyReplicasWithNoProgress = new ArrayList<>();
```

Comment:
rename to standByReplicas

Reply:
We are specifically wanting standby no progress replicas, so kept name here.